hist/histv7/inc/ROOT/RHistImpl.hxx

```diff
@@ -227,122 +257,467 @@ namespace Internal {
+      // we don't need to look at the remaining axes. Projecting on those
+      // dimensions would only take us into an hyperplane of over/underflow
+      // bins for the current axis, and we know that by construction there
+      // will be no regular bins in there.
+      if (localBins[I] < 1)
+         return total_regular_bins_before;
+      
+      return RComputeGlobalBin<I - 1, NDIMS, BINS, AXES>()(total_regular_bins_before, axes, virtualBins, binSizes, localBins);
+   }
+};
+
+/// Recursively compute some quantities needed for `ComputeLocalBins`, namely
+/// the total number of bins per hyperplane (overflow and regular) and the
+/// number of regular bins per hyperplane on the hyperplanes that have them.
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation;
+
+template <int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation<0, NDIMS, AXES> {
+   void operator()(std::array<int, NDIMS-1> /* bins_per_hyperplane */, std::array<int, NDIMS-1> /* regular_bins_per_hyperplane */, const AXES & /*axes*/) const
+   {}
+};
+
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBinsInitialisation {
+   void operator()(std::array<int, NDIMS-1>& bins_per_hyperplane, std::array<int, NDIMS-1>& regular_bins_per_hyperplane, const AXES &axes) const
+   {
+      constexpr const int thisAxis = NDIMS - I - 1;
+      bins_per_hyperplane[thisAxis] = Internal::RGetNBinsCount<thisAxis, AXES>()(axes);
+      regular_bins_per_hyperplane[thisAxis] = Internal::RGetNBinsNoOverCount<thisAxis, AXES>()(axes);
+      RComputeLocalBinsInitialisation<I - 1, NDIMS, AXES>()(bins_per_hyperplane, regular_bins_per_hyperplane, axes);
+   }
+};
+
+/// Recursively computes the number of regular bins before the current dimension,
+/// as well as the number of under- and overflow bins left to account for after
+/// the current dimension.
+/// If the latter is equal to 0, there is no need to process further.
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBins;
+
+template <int NDIMS, class AXES>
+struct RComputeLocalBins<0, NDIMS, AXES> {
+   std::array<int, 2> operator()(const AXES &/*axes*/, std::array<int, 2> parameters_to_process,
+         std::array<int, NDIMS-1> /* bins_per_hyperplane */, std::array<int, NDIMS-1> /* regular_bins_per_hyperplane */,
+         int /* curr_bins_per_hyperplane */, int /* curr_regular_bins_per_hyperplane */) const
+   {
+      return parameters_to_process;
+   }
+};
+
+template <int I, int NDIMS, class AXES>
+struct RComputeLocalBins {
+   std::array<int, 2> operator()(const AXES &axes, std::array<int, 2> parameters_to_process,
+         std::array<int, NDIMS-1> bins_per_hyperplane, std::array<int, NDIMS-1> regular_bins_per_hyperplane,
+         int curr_bins_per_hyperplane, int curr_regular_bins_per_hyperplane) const
+   {
+      int unprocessed_previous_overflow_bin = parameters_to_process[0];
+      int num_regular_bins_before = parameters_to_process[1];
```

Comment:
As an alternative to the `parameters_to_process` array indirection, considering passing the parameters by mutable reference:

```
int& unprocessed_previous_overflow_bin,
int& num_regular_bins_before,
```